csrc/predicate_compute.cpp

```diff
@@ -346,20 +355,30 @@ Val* PredicateCompute::getInlinePredicate(
     thread_pred = gpu_lower->kernel()->trueVal();
     // If it is a initilization op, return immediately.
     if (ir_utils::isTensorScalarFillOp(expr)) {
-      return thread_pred;
+      RECORD_AND_RETURN(thread_pred);
     }
   }
 
   if (loops.empty()) {
     NVF_ERROR(thread_pred != nullptr);
-    return thread_pred;
+    RECORD_AND_RETURN(thread_pred);
   }
 
   auto out_tv = ir_utils::getTvOutput(expr);
   NVF_ERROR(out_tv != nullptr, "Missing TensorView output");
 
   if (gpu_lower->predicateElimination().canOmitPredicate(expr)) {
-    return thread_pred;
+    RECORD_AND_RETURN(thread_pred);
+  }
+
+  auto parallel_dom_pred =
+      ParallelizedDomainPredicate::getPredicate(expr, loops);
+  NVF_ERROR(parallel_dom_pred != nullptr);
+
+  // TMA handles out-of-bounds accesses in hardware, so parallel_dom_pred
+  // itself is sufficient to predicate the accesses.
+  if (ir_utils::isCpAsyncBulk(expr)) {
```

Comment:
A side/silly question - having predicated returned for LoadStoreOp set to bulk will also create a IfElseThen with predicate for all the Exprs created around it? For example for mbarrier:arriveExpectTx and wait, like in the examples?

Reply:
Here, only this particular `expr` is predicated, but when this `expr` is lowered, IIUC, it's translated to multiple expressions, including the arrive and wait. Am I correct, @zasdfgbnm?

Reply:
Yes, what @naoyam said is correct.